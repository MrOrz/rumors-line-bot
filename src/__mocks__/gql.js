const mockResultQueue = [];

function gqlMock() {
  return () => {
    const ret = mockResultQueue.shift();
    return Promise.resolve(ret);
  };
}

/**
 * Enqueues mock data what will be returned in next gql()() call.
 *
 * Once gql()() is invoked, the first data in queue will be consumed.
 *
 * @param {*} returnValue the mock value that gql()()'s returned promise will resolve to.
 * @returns nothing
 */
gqlMock.__push = function(returnValue) {
  mockResultQueue.push(returnValue);
};

export default gqlMock;
