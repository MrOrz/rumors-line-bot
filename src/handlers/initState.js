import stringSimilarity from 'string-similarity';
import gql from '../gql';
import { createPostbackAction } from './utils';

const SIMILARITY_THRESHOLD = 0.95;

export default async function initState(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  // Store user input into context
  data.searchedText = event.input;

  // Search for articles
  const { data: { ListArticles } } = await gql`
    query($text: String!) {
      ListArticles(
        filter: { moreLikeThis: { like: $text } }
        orderBy: [{ _score: DESC }]
        first: 4
      ) {
        edges {
          node {
            text
            id
          }
        }
      }
    }
  `({
    text: event.input,
  });

  const articleSummary = `${event.input.slice(0, 10)}${event.input.length > 10 ? '⋯⋯' : ''}`;

  if (ListArticles.edges.length) {
    const edgesSortedWithSimilarity = ListArticles.edges
      .map(edge => {
        edge.similarity = stringSimilarity.compareTwoStrings(
          // Remove spaces so that we count word's similarities only
          //
          edge.node.text.replace(/\s/g, ''),
          event.input.replace(/\s/g, '')
        );
        return edge;
      })
      .sort((edge1, edge2) => edge2.similarity - edge1.similarity);

    // Store article ids
    data.foundArticleIds = edgesSortedWithSimilarity.map(
      ({ node: { id } }) => id
    );

    const hasIdenticalDocs =
      edgesSortedWithSimilarity[0].similarity >= SIMILARITY_THRESHOLD;

    if (edgesSortedWithSimilarity.length === 1 && hasIdenticalDocs) {
      // choose for user
      event.input = 1;

      return {
        data,
        state: 'CHOOSING_ARTICLE',
        event,
        issuedAt,
        userId,
        replies,
        isSkipUser: true,
      };
    }

    const templateMessage = {
      type: 'template',
      altText: edgesSortedWithSimilarity
        .map(
          ({ node: { text } }, idx) => `選擇請打 ${idx + 1}> ${text.slice(0, 20)}`
        )
        .concat(hasIdenticalDocs ? [] : ['若以上皆非，請打 0。'])
        .join('\n\n'),
      template: {
        type: 'carousel',
        columns: edgesSortedWithSimilarity
          .map(({ node: { text }, similarity }, idx) => ({
            text: `[相似度:${(similarity * 100).toFixed(2) + '%'}] \n ${text.slice(0, 100)}`,
            actions: [createPostbackAction('選擇此則', idx + 1, issuedAt)],
          }))
          .concat(
            hasIdenticalDocs
              ? []
              : [
                  {
                    text: '這裡沒有一篇是我傳的訊息。',
                    actions: [createPostbackAction('選擇', 0, issuedAt)],
                  },
                ]
          ),
      },
    };

    replies = [
      {
        type: 'text',
        text: `幫您查詢「${articleSummary}」的相關回應。`,
      },
      {
        type: 'text',
        text: '請問下列文章中，哪一篇是您剛才傳送的訊息呢？',
      },
      templateMessage,
    ];
    state = 'CHOOSING_ARTICLE';
  } else {
    replies = [
      {
        type: 'text',
        text: `找不到關於「${articleSummary}」文章耶 QQ`,
      },
      {
        type: 'template',
        altText: '請問要將這份文章送出到資料庫嗎？\n「是」請輸入「y」，「否」請輸入「n」或其他單一字母。',
        template: {
          type: 'buttons',
          text: '請問要將這份文章送出到資料庫嗎？',
          actions: [
            createPostbackAction('是', 'y', issuedAt),
            createPostbackAction('否', 'n', issuedAt),
          ],
        },
      },
    ];
    state = 'ASKING_ARTICLE_SUBMISSION';
  }
  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
