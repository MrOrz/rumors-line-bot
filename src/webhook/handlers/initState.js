import stringSimilarity from 'string-similarity';
import { t } from 'ttag';
import gql from 'src/lib/gql';
import {
  createPostbackAction,
  ellipsis,
  createAskArticleSubmissionConsent,
} from './utils';
import ga from 'src/lib/ga';

const SIMILARITY_THRESHOLD = 0.95;

export default async function initState(params) {
  let { data, state, event, userId, replies, isSkipUser } = params;

  // Track text message type send by user
  const visitor = ga(userId, state, event.input);
  visitor.event({ ec: 'UserInput', ea: 'MessageType', el: 'text' });

  // Store user input into context
  data.searchedText = event.input;
  // Store input message type to context for non-init states use
  data.messageType = event.message.type;

  // Search for articles
  const {
    data: { ListArticles },
  } = await gql`
    query($text: String!) {
      ListArticles(
        filter: { moreLikeThis: { like: $text } }
        orderBy: [{ _score: DESC }]
        first: 4
      ) {
        edges {
          node {
            text
            id
          }
        }
      }
    }
  `({
    text: event.input,
  });

  const articleSummary = ellipsis(event.input, 12);

  if (ListArticles.edges.length) {
    // Track if find similar Articles in DB.
    visitor.event({ ec: 'UserInput', ea: 'ArticleSearch', el: 'ArticleFound' });

    // Track which Article is searched. And set tracking event as non-interactionHit.
    ListArticles.edges.forEach(edge => {
      visitor.event({
        ec: 'Article',
        ea: 'Search',
        el: edge.node.id,
        ni: true,
      });
    });

    const edgesSortedWithSimilarity = ListArticles.edges
      .map(edge => {
        edge.similarity = stringSimilarity.compareTwoStrings(
          // Remove spaces so that we count word's similarities only
          //
          edge.node.text.replace(/\s/g, ''),
          event.input.replace(/\s/g, '')
        );
        return edge;
      })
      .sort((edge1, edge2) => edge2.similarity - edge1.similarity)
      .slice(0, 9) /* flex carousel has at most 10 bubbles */;

    // Store article ids
    data.foundArticleIds = edgesSortedWithSimilarity.map(
      ({ node: { id } }) => id
    );

    const hasIdenticalDocs =
      edgesSortedWithSimilarity[0].similarity >= SIMILARITY_THRESHOLD;

    if (edgesSortedWithSimilarity.length === 1 && hasIdenticalDocs) {
      // choose for user
      event.input = 1;

      visitor.send();
      return {
        data,
        state: 'CHOOSING_ARTICLE',
        event,
        userId,
        replies,
        isSkipUser: true,
      };
    }

    const postMessage = edgesSortedWithSimilarity.map(
      ({ node: { text }, similarity }, idx) => {
        const similarityPercentage = Math.round(similarity * 100);
        const similarityEmoji = ['üòê', 'üôÇ', 'üòÄ', 'üòÉ', 'üòÑ'][
          Math.floor(similarity * 4.999)
        ];
        return {
          type: 'bubble',
          direction: 'ltr',
          header: {
            type: 'box',
            layout: 'horizontal',
            spacing: 'md',
            paddingBottom: 'none',
            contents: [
              {
                type: 'text',
                text: similarityEmoji,
                flex: 0,
              },
              {
                type: 'text',
                text: t`Looks ${similarityPercentage}% similar`,
                gravity: 'center',
                size: 'sm',
                weight: 'bold',
                wrap: true,
                color: '#AAAAAA',
              },
            ],
          },
          body: {
            type: 'box',
            layout: 'horizontal',
            spacing: 'none',
            margin: 'none',
            contents: [
              {
                type: 'text',
                text: ellipsis(text, 300, '...'), // 50KB for entire Flex carousel,
                maxLines: 6,
                flex: 0,
                gravity: 'top',
                weight: 'regular',
                wrap: true,
              },
            ],
          },
          footer: {
            type: 'box',
            layout: 'horizontal',
            contents: [
              {
                type: 'button',
                action: createPostbackAction(
                  t`Choose this one`,
                  idx + 1,
                  data.sessionId
                ),
                style: 'primary',
              },
            ],
          },
        };
      }
    );
    postMessage.push({
      type: 'bubble',
      header: {
        type: 'box',
        layout: 'horizontal',
        paddingBottom: 'none',
        contents: [
          {
            type: 'text',
            text: 'üò∂',
            margin: 'none',
            size: 'sm',
            weight: 'bold',
            color: '#AAAAAA',
          },
        ],
      },
      body: {
        type: 'box',
        layout: 'horizontal',
        spacing: 'none',
        margin: 'none',
        contents: [
          {
            type: 'text',
            text: t`None of these messages matches mine :(`,
            maxLines: 5,
            flex: 0,
            gravity: 'top',
            weight: 'regular',
            wrap: true,
          },
        ],
      },
      footer: {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'button',
            action: createPostbackAction(t`Tell us more`, 0, data.sessionId),
            style: 'primary',
          },
        ],
      },
    });

    const templateMessage = {
      type: 'flex',
      altText: t`Please choose the most similar message from the list.`,
      contents: {
        type: 'carousel',
        contents: postMessage,
      },
    };

    const prefixTextArticleFound = [
      {
        type: 'text',
        text: `üîç ${t`There are some messages that looks similar to "${articleSummary}" you have sent to me.`}`,
      },
    ];
    const textArticleFound = [
      {
        type: 'text',
        text:
          t`Internet rumors are often mutated and shared.
            Please choose the version that looks the most similar` + 'üëá',
      },
      templateMessage,
    ];
    if (data.messageType === 'image') {
      replies = textArticleFound;
    } else {
      replies = prefixTextArticleFound.concat(textArticleFound);
    }

    state = 'CHOOSING_ARTICLE';
  } else {
    // Track if find similar Articles in DB.
    visitor.event({
      ec: 'UserInput',
      ea: 'ArticleSearch',
      el: 'ArticleNotFound',
    });

    replies = [
      {
        type: 'text',
        // use `articleSummary` for text only because ocr may get wrong text from image
        text:
          data.messageType === 'text'
            ? t`We didn't find anything about "${articleSummary}" :(`
            : t`We didn't find anything about that :(`,
      },
      ...createAskArticleSubmissionConsent(userId, data.sessionId),
    ];
    state = 'ASKING_ARTICLE_SUBMISSION_CONSENT';
  }
  visitor.send();
  return { data, state, event, userId, replies, isSkipUser };
}
