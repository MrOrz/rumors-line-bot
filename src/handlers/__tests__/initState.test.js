import initState from '../initState';
import * as gql from '../../gql';

describe('initState(params)', () => {
  it('article found', done => {
    const input = {
      data: {},
      state: '__INIT__',
      event: {
        type: 'message',
        input: '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入',
        timestamp: 1497994016356,
        message: {
          type: 'text',
          id: '6270464463537',
          text: '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入',
        },
      },
      issuedAt: 1497994017447,
      userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
      replies: undefined,
      isSkipUser: false,
    };

    /* eslint-disable import/namespace */
    gql.default = () => () => {
      return new Promise(resolve => {
        resolve({
          data: {
            ListArticles: {
              edges: [
                {
                  node: {
                    text: '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入，下車時使用老人悠遊卡，跳錶車資105元，優惠32元，只扣73元，哈哈，這是屬於我們的福利，與大家分享，可以善加利用！=7折，朋友使用ok',
                    id: 'AVvY-yizyCdS-nWhuYWx',
                  },
                },
              ],
            },
          },
        });
      });
    };

    initState(input).then(
      result => {
        expect(result).toMatchSnapshot();
        done();
      },
      error => {
        console.trace(error);
        done();
      }
    );
  });

  it('article found with 120 words limit', done => {
    const input = {
      data: {},
      state: '__INIT__',
      event: {
        type: 'message',
        input: '這樣的大事國內媒體竟然不敢報導！\n我國駐日代表將原「中華民國」申請更名為「台灣」結果被日本裁罰，須繳納7000萬日圓（合約台幣2100萬元）高額稅賦(轉載中時電子報）\n\n我駐日代表謝長廷將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭自認友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元）高額稅賦...民進黨沒想到如此更名竟然是這樣的下場：被他最信任也最友好的日本政府給坑了。\n果然錯誤的政策比貪污可怕，2100萬就這樣打水漂了，還要資助九州水患，核四停建違約賠償金.......夠全國軍公教退休2次.........\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118',
        timestamp: 1502477505309,
        message: {
          type: 'text',
          id: '6530038889933',
          text: '這樣的大事國內媒體竟然不敢報導！\n我國駐日代表將原「中華民國」申請更名為「台灣」結果被日本裁罰，須繳納7000萬日圓（合約台幣2100萬元）高額稅賦(轉載中時電子報）\n\n我駐日代表謝長廷將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭自認友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元）高額稅賦...民進黨沒想到如此更名竟然是這樣的下場：被他最信任也最友好的日本政府給坑了。\n果然錯誤的政策比貪污可怕，2100萬就這樣打水漂了，還要資助九州水患，核四停建違約賠償金.......夠全國軍公教退休2次.........\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118',
        },
      },
      issuedAt: 1502477506267,
      userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
      isSkipUser: false,
    };

    /* eslint-disable import/namespace */
    gql.default = () => () => {
      return Promise.resolve({
        data: {
          ListArticles: {
            edges: [
              {
                node: {
                  text: ' 這樣的大事國內媒體竟然不敢報導！\n我國駐日代表將原「中華民國」申請更名為「台灣」結果被日本裁罰，須繳納7000萬日圓（合約台幣2100萬元）高額稅賦(轉載中時電子報）\n\n我駐日代表謝長廷將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭自認友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元）高額稅賦...民進黨沒想到如此更名竟然是這樣的下場：被他最信任也最友好的日本政府給坑了。\n果然錯誤的政策比貪污可怕，2100萬就這樣打水漂了，還要資助九州水患，核四停建違約賠償金.......夠全國軍公教退休2次.........\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118',
                  id: 'AV00D5G-yCdS-nWhucFj',
                },
              },
              {
                node: {
                  text: '●我國駐日代表將原「中華民國」申請更名為~「台灣」，結果被日本國裁罰，須繳納7000萬日圓（合約台幣2100萬元…）的高額稅賦，(被封殺的電視新聞~這難道不是民脂民膏嗎…？)...轉載中時電子報\n\n我駐日代表-謝長廷，將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭…友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元…，這難到不是 人民的血汗錢嗎…？）高額稅賦...！\n民進黨沒想到，如此更名，竟然是這樣的下場：\n被他最信任，也最友好的日本政府給坑了，堂堂一個駐日代表謝長廷，竟然被日本乾爹給坑了，實在是無顏見江東父老…！ ......\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118\n\n',
                  id: 'AV0x1TlTyCdS-nWhucDm',
                },
              },
            ],
          },
        },
      });
    };

    initState(input).then(
      result => {
        for (let reply of result.replies) {
          if (reply.type === 'text') {
            expect(reply.text.length < 120).toBe(true);
          } else if (reply.type === 'template') {
            expect(reply.altText.length < 120).toBe(true);
            if (reply.template.type === 'carousel') {
              for (let column of reply.template.columns) {
                expect(column.text.length < 120).toBe(true);
              }
            }
          }
        }
        done();
      },
      error => {
        console.trace(error);
        done();
      }
    );
  });
  it('article not found');
});
