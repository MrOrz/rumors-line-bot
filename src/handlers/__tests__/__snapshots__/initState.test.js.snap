// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article found 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
    ],
    "searchedText": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
  },
  "event": Object {
    "input": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
    "message": Object {
      "id": "6270464463537",
      "text": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "text": "幫您查詢「計程車上有裝悠遊卡感⋯⋯」的相關回應。",
      "type": "text",
    },
    Object {
      "text": "請問下列文章中，哪一篇是您剛才傳送的訊息呢？",
      "type": "text",
    },
    Object {
      "altText": "選擇請打 1> 計程車上有裝悠遊卡感應器，老人悠悠卡可以

若以上皆非，請打 0。",
      "template": Object {
        "columns": Array [
          Object {
            "actions": Array [
              Object {
                "data": "{\\"input\\":1,\\"issuedAt\\":1497994017447}",
                "label": "選擇此則",
                "type": "postback",
              },
            ],
            "text": "[相似度:52.41%] 
 計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入，下車時使用老人悠遊卡，跳錶車資105元，優惠32元，只扣73元，哈哈，這是屬於我們的福利，與大家分享，可以善加利用！=7",
          },
          Object {
            "actions": Array [
              Object {
                "data": "{\\"input\\":0,\\"issuedAt\\":1497994017447}",
                "label": "選擇",
                "type": "postback",
              },
            ],
            "text": "這裡沒有一篇是我傳的訊息。",
          },
        ],
        "type": "carousel",
      },
      "type": "template",
    },
  ],
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`articles found with high similarity 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
      "AVvY-yizyCdS-nWhuYWy",
    ],
    "searchedText": "YouTube · 寻找健康人生",
  },
  "event": Object {
    "input": "YouTube · 寻找健康人生",
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "text": "幫您查詢「YouTube · ⋯⋯」的相關回應。",
      "type": "text",
    },
    Object {
      "text": "請問下列文章中，哪一篇是您剛才傳送的訊息呢？",
      "type": "text",
    },
    Object {
      "altText": "選擇請打 1> YouTube  ·  寻找健康人生

選擇請打 2> YouTube
·
寻找健康人生",
      "template": Object {
        "columns": Array [
          Object {
            "actions": Array [
              Object {
                "data": "{\\"input\\":1,\\"issuedAt\\":1497994017447}",
                "label": "選擇此則",
                "type": "postback",
              },
            ],
            "text": "[相似度:100.00%] 
 YouTube  ·  寻找健康人生",
          },
          Object {
            "actions": Array [
              Object {
                "data": "{\\"input\\":2,\\"issuedAt\\":1497994017447}",
                "label": "選擇此則",
                "type": "postback",
              },
            ],
            "text": "[相似度:100.00%] 
 YouTube
·
寻找健康人生",
          },
        ],
        "type": "carousel",
      },
      "type": "template",
    },
  ],
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`only one article found with high similarity 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
    ],
    "searchedText": "YouTube · 寻找健康人生",
  },
  "event": Object {
    "input": 1,
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": true,
  "issuedAt": 1497994017447,
  "replies": undefined,
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should handle article not found 1`] = `
Object {
  "data": Object {
    "searchedText": "YouTube · 寻找健康人生",
  },
  "event": Object {
    "input": "YouTube · 寻找健康人生",
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1497994017447,
  "replies": Array [
    Object {
      "text": "找不到關於「YouTube · ⋯⋯」訊息耶 QQ",
      "type": "text",
    },
    Object {
      "altText": "【送出訊息到公開資料庫？】
若這是「轉傳訊息」，而且您覺得這很可能是一則「謠言」，請將這則訊息送進公開資料庫建檔，讓好心人查證與回覆。
雖然您不會立刻收到查證結果，但可以幫助到未來同樣收到這份訊息的人。

「送出」請輸入「y」，「放棄」則請輸入「n」或其他單一字母。",
      "template": Object {
        "actions": Array [
          Object {
            "data": "{\\"input\\":\\"y\\",\\"issuedAt\\":1497994017447}",
            "label": "我要送出",
            "type": "postback",
          },
          Object {
            "data": "{\\"input\\":\\"n\\",\\"issuedAt\\":1497994017447}",
            "label": "放棄",
            "type": "postback",
          },
        ],
        "text": "【送出訊息到公開資料庫？】
若這是「轉傳訊息」，而且您覺得這很可能是一則「謠言」，請將這則訊息送進公開資料庫建檔，讓好心人查證與回覆。
雖然您不會立刻收到查證結果，但可以幫助到未來同樣收到這份訊息的人。
",
        "type": "buttons",
      },
      "type": "template",
    },
  ],
  "state": "ASKING_ARTICLE_SUBMISSION",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;
